exonum/src/storage/proof_map_index/key.rs

```diff
@@ -75,52 +69,117 @@ impl ::std::ops::Not for ChildKind {
         debug_assert_eq!(key.size(), KEY_SIZE);
 
-        let mut data = [0; KEY_SIZE];
-        key.write(&mut data);
-        DBKey {
-            data: data,
-            from: 0,
-            to: (KEY_SIZE * 8) as u16,
-        }
+        let mut data = [0; PROOF_PATH_SIZE];
+        data[0] = LEAF_KEY_PREFIX;
+        key.write(&mut data[1..KEY_SIZE + 1]);
+        data[PROOF_PATH_LEN_POS] = 0;
+        ProofPath::from_raw(data)
     }
 
-    #[doc(hidden)]
-    pub fn from(&self) -> u16 {
-        self.from
+    /// Shows the type of path.
+    pub fn is_leaf(&self) -> bool {
+        self.bytes[0] == LEAF_KEY_PREFIX
     }
 
-    // TODO: terrible hack, try to remove this (ECR-22)
-    #[doc(hidden)]
-    pub fn set_from(&mut self, from: u16) {
-        self.from = from
+    /// Returns the byte representation of `ProofPath`.
+    pub fn as_bytes(&self) -> &[u8] {
+        &self.bytes
     }
 
-    #[doc(hidden)]
-    pub fn to(&self) -> u16 {
-        self.to
+    /// Constructs the `ProofPath` from raw bytes.
+    fn from_raw(raw: [u8; PROOF_PATH_SIZE]) -> ProofPath {
+        debug_assert!(
+            if raw[PROOF_PATH_KIND_POS] == LEAF_KEY_PREFIX {
+                raw[PROOF_PATH_LEN_POS] == 0
+            } else {
+                true
+            },
+            "ProofPath is inconsistent"
+        );
+
+        ProofPath {
+            bytes: raw,
+            start: 0,
+        }
     }
 
-    /// Returns length of the `DBKey`.
-    pub fn len(&self) -> u16 {
-        self.to - self.from
+    /// Sets the right border of the bit range.
+    fn set_end(&mut self, end: u16) {
+        let max_len = (KEY_SIZE * 8) as u16;
+        assert!(end <= max_len);
+        // Update ProofPath kind and right bound.
+        if end == max_len {
+            self.bytes[0] = LEAF_KEY_PREFIX;
+            self.bytes[PROOF_PATH_LEN_POS] = 0;
+        } else {
+            self.bytes[0] = BRANCH_KEY_PREFIX;
+            self.bytes[PROOF_PATH_LEN_POS] = end as u8;
+        };
     }
+}
 
-    /// Returns true if `DBKey` has zero length.
-    pub fn is_empty(&self) -> bool {
-        self.to == self.from
+/// The bits representation of the `ProofPath`.
+pub trait BitsRange {
+    /// Returns the left border of the range.
+    fn start(&self) -> u16;
+    /// Returns the right border of the range.
+    fn end(&self) -> u16;
+    /// Returns length in bits of the range.
+    fn len(&self) -> u16 {
+        self.end() - self.start()
+    }
+    /// Returns true if the range has zero length.
+    fn is_empty(&self) -> bool {
+        self.end() == self.start()
     }
-
     /// Get bit at position `idx`.
-    pub fn get(&self, idx: u16) -> ChildKind {
-        debug_assert!(self.from + idx < self.to);
+    fn bit(&self, idx: u16) -> ChildKind;
+    /// Returns the new `ProofPath` with the given left border.
+    fn start_from(&self, idx: u16) -> Self;
+    /// Shortens this ProofPath to the specified length.
```

Comment:
Nitpick: "shortens" sounds like the transformation is performed in-place; propose changing to something like "Returns a copy of this bit range shortened to the specified length".